src/router/Router.js

```diff
@@ -279,56 +291,63 @@ export default class Router extends DisposableBase {
+                                if (!modelRecord.hasChanges && wasDispatched) {
+                                    modelRecord.hasChanges = true;
+                                }
+                                hasEvents = modelRecord.eventQueue.length > 0;
+                                if (hasEvents) {
+                                    eventRecord = modelRecord.eventQueue.shift();
+                                }
+                            } // keep looping until any events from the dispatch to processors stage are processed
+                            this._diagnosticMonitor.finishDispatchingEvent();
+                            if (!modelRecord.wasRemoved) {
+                                this._diagnosticMonitor.postProcessingModel();
+                                this._state.moveToPostProcessing();
+                                modelRecord.runPostEventProcessor(modelRecord.model);
+                                if (modelRecord.model.lock && typeof modelRecord.model.lock === 'function') {
+                                    modelRecord.model.lock();
+                                }
                             }
                         }
-                    }
-                    this.broadcastEvent(Const.modelChangedEvent, new ModelChangedEvent(modelRecord.modelId, modelRecord.model));
+                        this.broadcastEvent(Const.modelChangedEvent, new ModelChangedEvent(modelRecord.modelId, modelRecord.model));
+                        modelRecord = this._getNextModelRecordWithQueuedEvents();
+                        hasEvents = typeof modelRecord !== 'undefined';
+                        this._diagnosticMonitor.endingModelEventLoop();
+                    }  // keep looping until any events raised during post event processing OR event that have come in for other models are processed
+                    this._state.moveToDispatchModelUpdates();
+                    this._dispatchModelUpdates();
                     modelRecord = this._getNextModelRecordWithQueuedEvents();
                     hasEvents = typeof modelRecord !== 'undefined';
-                    this._diagnosticMonitor.endingModelEventLoop();
-                }  // keep looping until any events raised during post event processing OR event that have come in for other models are processed
-                this._state.moveToDispatchModelUpdates();
-                this._dispatchModelUpdates();
-                modelRecord = this._getNextModelRecordWithQueuedEvents();
-                hasEvents = typeof modelRecord !== 'undefined';
-            } // keep looping until any events from the dispatch updates stages are processed
+                } // keep looping until any events from the dispatch updates stages are processed
+            }
+            catch(err) {
+                _log.error('Router saw an error purging queues: [{0}]', err);
+                this._onErrorHandler(err);
```

Comment:
I think this try catch and dispatch is in the wrong place. There is existing halting flows we should place the dispatch to the error handler there. The `_halt(err)` is the place to put it. If you want to distinguish between the initial exception and follow-on exceptions we can do that within the `_halt()` method. For example any interaction with the router after halting will results in further calls to `_halt()`, that method could query `_state` and determine if it's the initial exception or a follow-on.